File: iosApp/iosApp/inlineclasses/InlineClassesScreenViewModel.swift
//
//  InlineClassesScreenViewModel.swift
//  iosApp
//
//  Created by Emanuele on 16/09/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import shared
import sharedSwift

class InlineClassesScreenViewModel: ObservableObject {
    
    @Published var state: InlineClassesScreenState = InlineClassesScreenState()
    
    func onShowMyNameClicked() {
        state.formattedName = ProfileFormatterKt.formatFirstAndLastName(firstName: FirstNameIos(firstName: "John"), lastName_: LastNameIos(lastName: "Doe"))
    }
    
    func onShowMyNameUsingInlineClassesClicked() {
        //Inline classes are converted to Any type, so any value can be passed as argument
        state.formattedNameUsingInlineClasses = ProfileFormatterKt.formatFirstAndLastName(firstName: "John", lastName: 123)
    }
    
}

File: androidApp/src/main/java/dev/emanuelepapa/kmmsurvivalguidesamples/android/inlineclasses/InlineClassesScreenViewModel.kt
package dev.emanuelepapa.kmmsurvivalguidesamples.android.inlineclasses

import androidx.lifecycle.ViewModel
import dev.emanuelepapa.kmmsurvivalguidesamples.inline.FirstName
import dev.emanuelepapa.kmmsurvivalguidesamples.inline.LastName
import dev.emanuelepapa.kmmsurvivalguidesamples.inline.formatFirstAndLastName
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

class InlineClassesScreenViewModel : ViewModel() {

    private val mutableState = MutableStateFlow(
        InlineClassesScreenState(
            formattedName = null
        )
    )
    val state: StateFlow<InlineClassesScreenState> = mutableState

    fun onShowMyNameClicked() {
        mutableState.value = mutableState.value.copy(
            formattedName = formatFirstAndLastName(
                FirstName("John"),
                LastName("Doe")
            )
        )
    }

}